internal/providers/ionoscloud/ionoscloud.go

```diff
@@ -0,0 +1,208 @@
+	defaultDeviceLabel    = "OEM"
+	userDataKernelFlag    = "ignition.config.path"
+	defaultUserDataPath   = "config/user-data"
+)
+
+func init() {
+	platform.Register(platform.Provider{
+		Name:  "ionoscloud",
+		Fetch: fetchConfig,
+	})
+}
+
+func fetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
+	var data []byte
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+
+	dispatch := func(name string, fn func() ([]byte, error)) {
+		raw, err := fn()
+		if err != nil {
+			switch err {
+			case context.Canceled:
+			case context.DeadlineExceeded:
+				f.Logger.Err("timed out while fetching config from %s", name)
+			default:
+				f.Logger.Err("failed to fetch config from %s: %v", name, err)
+			}
+			return
+		}
+
+		data = raw
+		cancel()
+	}
+
+	deviceLabel, userDataPath, err := readFromKernelParams(f.Logger)
+
+	if err != nil {
+		f.Logger.Err("couldn't read kernel parameters: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+
+	if deviceLabel == "" {
+		deviceLabel = defaultDeviceLabel
+	}
+
+	if userDataPath == "" {
+		userDataPath = defaultUserDataPath
+	}
+
+	go dispatch(
+		"load config from disk", func() ([]byte, error) {
+			return fetchConfigFromDevice(f.Logger, ctx, deviceLabel, userDataPath)
+		},
+	)
+
+	<-ctx.Done()
+	if ctx.Err() == context.DeadlineExceeded {
+		f.Logger.Info("disk was not available in time. Continuing without a config...")
+	}
+
+	return util.ParseConfig(f.Logger, data)
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return (err == nil)
+}
+
+func fetchConfigFromDevice(logger *log.Logger,
+	ctx context.Context,
+	deviceLabel string,
+	dataPath string,
+) ([]byte, error) {
+	device := filepath.Join(distro.DiskByLabelDir(), deviceLabel)
+	for !fileExists(device) {
+		logger.Debug("disk (%q) not found. Waiting...", device)
+		select {
+		case <-time.After(time.Second):
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+
+	logger.Debug("creating temporary mount point")
+	mnt, err := os.MkdirTemp("", "ignition-config")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp directory: %v", err)
+	}
+	defer os.Remove(mnt)
+
+	cmd := exec.Command(distro.MountCmd(), "-o", "ro", "-t", "auto", device, mnt)
+	if _, err := logger.LogCmd(cmd, "mounting disk"); err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = logger.LogOp(
+			func() error {
+				return ut.UmountPath(mnt)
+			},
+			"unmounting %q at %q", device, mnt,
+		)
+	}()
+
+	if !fileExists(filepath.Join(mnt, dataPath)) {
+		return nil, nil
+	}
+
+	contents, err := os.ReadFile(filepath.Join(mnt, dataPath))
+	if err != nil {
+		return nil, err
+	}
+
+	if util.IsCloudConfig(contents) {
+		logger.Debug("disk (%q) contains a cloud-config configuration, ignoring", device)
+		return nil, nil
+	}
+
+	if util.IsShellScript(contents) {
+		logger.Debug("disk (%q) contains a shell script, ignoring", device)
+		return nil, nil
+	}
```

Comment:
Can you elaborate on why we need this?

Reply:
Backwards compatibility for the Flatcar [CoreOS CloudInit fork](https://github.com/flatcar/coreos-cloudinit). But to be honest we don't really need it and can get rid of it.